homeassistant/components/mazda/climate.py

```diff
@@ -0,0 +1,197 @@
+
+    @property
+    def hvac_mode(self) -> HVACMode:
+        """Return the current HVAC setting."""
+        hvac_on = self.client.get_assumed_hvac_mode(self.vehicle_id)
+        if hvac_on:
+            return HVACMode.HEAT_COOL
+        return HVACMode.OFF
+
+    @property
+    def precision(self) -> float:
+        """Return the precision of the temperature setting."""
+        if self.data["hvacSetting"]["temperatureUnit"] == "F":
+            return PRECISION_WHOLE
+        return PRECISION_HALVES
+
+    @property
+    def current_temperature(self) -> float:
+        """Return the current temperature."""
+        current_temperature_celsius = self.data["evStatus"]["hvacInfo"][
+            "interiorTemperatureCelsius"
+        ]
+        if self.data["hvacSetting"]["temperatureUnit"] == "F":
+            return convert_temperature(
+                current_temperature_celsius, TEMP_CELSIUS, TEMP_FAHRENHEIT
+            )
+        return current_temperature_celsius
+
+    @property
+    def target_temperature(self) -> float:
+        """Return the temperature we try to reach."""
+        hvac_setting = self.client.get_assumed_hvac_setting(self.vehicle_id)
+
+        return hvac_setting.get("temperature")
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the unit of measurement."""
+        if self.data["hvacSetting"]["temperatureUnit"] == "F":
+            return TEMP_FAHRENHEIT
+        return TEMP_CELSIUS
+
+    @property
+    def min_temp(self) -> float:
+        """Return the minimum target temperature."""
+        if self.data["hvacSetting"]["temperatureUnit"] == "F":
+            return 61.0
+        if self.region == "MJO":
+            return 18.5
+        return 15.5
+
+    @property
+    def max_temp(self) -> float:
+        """Return the maximum target temperature."""
+        if self.data["hvacSetting"]["temperatureUnit"] == "F":
+            return 83.0
+        if self.region == "MJO":
+            return 31.5
+        return 28.5
+
+    @property
+    def preset_mode(self) -> str:
+        """Return the current preset mode based on the state of the defrosters."""
+        hvac_setting = self.client.get_assumed_hvac_setting(self.vehicle_id)
+
+        front_defroster = hvac_setting.get("frontDefroster")
+        rear_defroster = hvac_setting.get("rearDefroster")
+        if front_defroster and rear_defroster:
+            return PRESET_DEFROSTER_FRONT_AND_REAR
+        if front_defroster:
+            return PRESET_DEFROSTER_FRONT
+        if rear_defroster:
+            return PRESET_DEFROSTER_REAR
+        return PRESET_DEFROSTER_OFF
+
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set a new HVAC mode."""
+        if hvac_mode == HVACMode.HEAT_COOL:
+            await self.client.turn_on_hvac(self.vehicle_id)
+        elif hvac_mode == HVACMode.OFF:
+            await self.client.turn_off_hvac(self.vehicle_id)
+
+        self.async_write_ha_state()
```

Comment:
Is this needed here? We're not directly updating any state, and if we're counting on the coordinator to update as a result of turning on/off, then there's no need for this call.

Reply:
I included this so that HA would call `hvac_mode()` to get the updated HVAC mode after it was changed. If I remove this, then when I change the HVAC mode, the new HVAC mode is not reflected in HA for a while. It takes some time for the API response to update after a change.

Reply:
I see.
But what does `self.client.get_assumed_hvac_mode` do? The coordinator should be fetching new state, and properties should just be reading it.

Reply:
`self.client.get_assumed_hvac_mode` returns an "assumed" state value. Normally it just returns the HVAC mode from the mode recent API response. But if the HVAC mode was recently changed, it will return the new HVAC mode, even if that hasn't been returned in an API response yet.

When the HVAC mode is changed in HA, the coordinator doesn't seem to run its update function, and the properties don't read a new value. So that's why I need `self.async_write_ha_state()`, to make the properties read a new value from `self.client.get_assumed_hvac_mode()` that will reflect the newly updated setting. The lock entity in this integration works the same way.

Reply:
so `self.client.get_assumed_hvac_mode` is a cached value? It's not entirely clear from the code.
In any case, the correct way is probably to call on the coordinator to update again.

Reply:
> so `self.client.get_assumed_hvac_mode` is a cached value? It's not entirely clear from the code.

Yes, it's a cached value.

> In any case, the correct way is probably to call on the coordinator to update again.

The Mazda API has rate limiting, so I don't want to trigger a coordinator update every time the user changes the HVAC mode or target temperature. To avoid rate limiting issues I only want the coordinator to update every 3 minutes as scheduled.

Reply:
This still seems off to me to decouple fetching the new value and updating HA like that. It's confusing to see `self.async_write_ha_state()` in a function that doesn't directly update the state.
My 2c - I would store `hvac_mode` in the entity, and return it from the property, rather than call the library. I would then update the stored value both from the coordinator, and from this function.

Reply:
The problem is that there is some complex logic needed in order to determine whether the value from the API or the assumed value should be used. Right now this logic is located in the pymazda library, but if I followed your suggestion, I would need to move it from the library into the integration. In my PR where I added the lock entity, I actually had the logic in the integration, but was asked by a reviewer to move it out of the integration and into the library. See https://github.com/home-assistant/core/pull/50548#discussion_r631556364

Reply:
The logic would still stay in the library, but instead of being called from `hvac_mode()`, it will be called in the coordinator callback, and here.
So, for example, here it would be:
```python
self._hvac_mode = self.client.get_assumed_hvac_mode()
self.async_write_ha_state()
```

And then in the property you'll simply `return self._hvac_mode`

Reply:
OK I think I'm starting to see what you mean now. But I'm still confused about one thing. How would I update the `_hvac_mode` in the entity from the coordinator callback? I don't have a reference to the climate entity in the coordinator callback.

Reply:
You can add something like this to the entity:
```python
async def async_added_to_hass(self):
        """When entity is added to hass."""
        self.async_on_remove(
            self.coordinator.async_add_listener(self._refresh_from_coordinator)
        )
```

And then update everything you need in your implementation of `_refresh_from_coordinator`

Reply:
@OnFreund Thank you for the guidance. I had some time to take a look at this today, but would you mind explaining the above example in a bit more detail? I don't understand why `async_on_remove` is being used here. Wouldn't that only add the coordinator listener when the entity is removed? Why would we only want to add the listener if the entity is removed?

Reply:
The return value from `async_add_listener` is a function to call in order to remove the listener, and that's what you're passing to `async_on_remove()`, so that the listener can be removed when the entity is removed from HA.
You can think about it as:
```Python
remove_listener_func = self.coordinator.async_add_listener(self._refresh_from_coordinator)
self.async_on_remove(remove_listener_func)
```

Reply:
I've pushed an commit (https://github.com/home-assistant/core/pull/75037/commits/8100781748b130288037be702569a49f6c01b5f9) which implements your changes. Can you please take another look? Thanks!